.github/workflows/ci.yml

```diff
@@ -187,18 +187,42 @@ jobs:
     - name: Install
       run: |
         apt-get update -y
-        apt-get install -y ghc cabal-install
+        apt-get install -y autoconf build-essential zlib1g-dev libgmp-dev curl libncurses5 libtinfo5 libncurses5-dev libtinfo-dev
+        curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | BOOTSTRAP_HASKELL_NONINTERACTIVE=1 BOOTSTRAP_HASKELL_INSTALL_NO_STACK=1 sh
     - uses: actions/checkout@v1
-    # We need to break dependency circuit between optparse-applicative,
-    # process, unix and bytestring. Newer versions of Cabal are capable
-    # of coming up with these automatically, but older ones need a hint.
-    # Specifically, we need to inform Cabal that:
-    #
-    # * optparse-applicative doesn't use process;
-    # * QuickCheck uses the old random package; and
-    # * tasty doesn't use unix.
     - name: Test
       run: |
+        source ~/.ghcup/env
         cabal update
-        cabal install --dependencies-only --enable-tests --constraint 'optparse-applicative -process' --constraint 'QuickCheck +old-random' --constraint 'tasty -unix'
         cabal test
+
+  # We use github.com/haskell self-hosted runners for ARM testing.
+  # If they become unavailable in future, put ['armv7', 'aarch64']
+  # back to emulation jobs above.
+  arm:
+    needs: build
+    runs-on: [self-hosted, Linux, ARM64]
```

Comment:
Only one of the runners can execute 32bit.

Reply:
Are you sure? I have not spot any issues so far using both runners.

Reply:
Ok @angerman says the aarch64-linux-2-1-new machine can do EL0 AArch32 (32bit applications), just no full system AArch32 emulation.